### PR TITLE
[cowprotocol] Add surplus usd amount to trades table

### PR DESCRIPTION
**Pull request by fleupold**
Adding a common field which can be derived from the existing trade logic but involves some math and is therefore easy to mess up. Making this a field of the materialized view will make our dashboard simpler.

cc @bh2smith @gentrexha 

**For Dune Engine V2**

I've checked that:
* [ x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x ] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [ x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`

### CODE REVIEW

File: models/cow_protocol/ethereum/cow_protocol_ethereum_trades.sql [hunk 1]
```diff
@@ -226,4 +226,6 @@ valued_trades as (
         ON trades.order_uid = efs.order_uid
 )
 
-select * from valued_trades
+select *,
+  ((limit_sell_amount - atoms_sold) / atoms_sold  * usd_value) + ((atoms_bought - limit_buy_amount) / atoms_bought * usd_value) as surplus_usd
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_trades.sql:230]: Do you have a link to a query showing that this works? I thought we would have to perform the calculation differently depending on whether it is a buy or sell order. I assume that one of these two terms is zero and that is why there is no need for a case statement (could we put a comment here for that).

**Review by bh2smith — APPROVED** [review]
Looks good. Would like to see a comment about why there is no case statement (assuming some mathematical properties of the formula).

**Comment by fleupold** [comment]
Looks like `dbt compile` is failing because `surplus_usd` already exists as a column name on another table (which strikes me as an odd requirement).

Shouldn't be a problem however, as long as @gentrexha merges his changes that rename this column on the price improvement view first.

>  Error reading spellbook: cow_protocol/ethereum/cow_protocol_ethereum_schema.yml - Runtime Error
    Syntax error near line 305
    ------------------------------
    302|         name: surplus_atoms
    303|       - &surplus_percentage
    304|         name: surplus_percentage
    305|       - &surplus_usd
    306|         name: surplus_usd
    Raw Error:
    ------------------------------
    found duplicate anchor; first occurrence
      in "<unicode string>", line 132, column 9
    second occurrence
      in "<unicode string>", line 305, column 9

**Comment by bh2smith** [comment]
There is a star notation used to indicate and a column name that already exists. I guess it's so to avoid duplicate descriptions. You can use *surplus_usd without any description and it will be happy again.

**Comment by bh2smith** [comment]
It's probably the price improvement table where surplus usd is appearing twice. If you rebase this on @gentrexha 's PR from yesterday, he has already renamed this field and you should have no problems.

**Comment by bh2smith** [comment]
That is, we should rebase this on #2781 (which I will do for you now).

File: models/cow_protocol/ethereum/cow_protocol_ethereum_trades.sql [hunk 2]
```diff
@@ -226,4 +226,8 @@ valued_trades as (
         ON trades.order_uid = efs.order_uid
 )
 
-select * from valued_trades
+select *,
+  -- Depending on the order kind (buy vs sell), the limit amount will always be equal to either the amount bought or sold.
+  -- We can therefore sum up the surplus relative to total buy and sell amount and extrapolate given the total trade's usd value.
+  ((limit_sell_amount - atoms_sold) / atoms_sold  * usd_value) + ((atoms_bought - limit_buy_amount) / atoms_bought * usd_value) as surplus_usd
```
Comment (bh2smith) [inline models/cow_protocol/ethereum/cow_protocol_ethereum_trades.sql:232]: I wonder what happens here for partially-fillable orders... also couldn't we factor the `* usd_value` out of both terms here?
```suggestion
  usd_value * (((limit_sell_amount - atoms_sold) / atoms_sold) + ((atoms_bought - limit_buy_amount) / atoms_bought)) as surplus_usd
```

Please feel free to ignore the suggestion

**Review by bh2smith — APPROVED** [review]
I am really looking forward to this change!

**Comment by fleupold** [comment]
> I wonder what happens here for partially-fillable orders...

That's a good point, it wouldn't work with the current query. I've rewritten it to be based on the difference of ratios of limit and executed price (I made sure for existing orders the computations yield the same results module double imprecision): https://dune.com/queries/2044305

File: models/cow_protocol/gnosis/cow_protocol_gnosis_trades.sql [hunk 3]
```diff
@@ -212,4 +212,8 @@ valued_trades as (
                   ON uid = order_uid
 )
 
-select * from valued_trades
+select *,
+  -- Relative surplus (in %) is the difference between limit price and executed price as a ratio of the limit price.
+  -- Absolute surplus (in USD) is relative surplus ormer multiplied with the value of the trade
```
Comment (bh2smith) [inline models/cow_protocol/gnosis/cow_protocol_gnosis_trades.sql:217]: I can't quite decipher what "ormer" is here. Best guess is "former"...

**Comment by gentrexha** [comment]
Bumping this to get this merged

cc @jeff-dude @Hosuke

**Comment by jeff-dude** [comment]
@gentrexha as hosuke mentioned, please do look into merge conflicts here now that #2781 was merged and they had similar files edited.

**Comment by gentrexha** [comment]
Tagging @fleupold because I don't have access to the branch.

**Review by Hosuke — APPROVED** [review]
LGTM.✅
This PR adds a new column `surplus_usd` to `cow_protocol_[ ethereum | gnosis ]_trades`.
Thank you @fleupold 

[Sample cow_protocol ethereum trades with surplus_usd](https://dune.com/queries/2074283)
[Sample cow_protocol gnosis trades with surplus_usd](https://dune.com/queries/2074174?d=1)

**Review by jeff-dude — APPROVED** [review]
thanks all

**Pull request merged**